Add CallComputedMethod operation

amarekano opened the pull request:
This PR adds support to generating computed method calls to fuzzilli. This would now allow generating JS of the form:
 ```
let v = callee[method](arguments)
```
An example of this usecase would be the use of `Symbol.iterator` demonstrated in this [JSC testcase](https://github.com/WebKit/WebKit/blob/main/JSTests/stress/arguments-iterator.js#L36)

saelo commented on Sources/Fuzzilli/Core/CodeGenerators.swift:
```diff
@@ -248,6 +248,16 @@ public let CodeGenerators: [CodeGenerator] = [
+            methodName = b.genMethodName()
+        }
+        guard let arguments = b.randCallArguments(forMethod: methodName!, on: obj) else { return }
+        b.callMethod(methodName!, on: obj, withArgs: arguments)
```
shouldn't this be `callComputedMethod`?

saelo reviewed (commented):
Nice, thanks for implementing this!
So in general, I think the `CallComputedMethod` should be similar to e.g. `LoadComputedProperty` in that it uses a variable as method name and not a string constant. Then, the ProgramBuilder interface would be `public func callComputedMethod(_ name: Variable, on object: Variable, withArgs arguments: [Variable]) -> Variable`.
For the CodeGenerator, I think what you could do is basically this:
```
CodeGenerator("ComputedMethodCallGenerator", input: .object()) { b, obj in
    var methodName = b.type(of: obj).randomMethod()
    if methodName == nil { ... }
    let method = b.loadString(methodName)
    guard let arguments = b.randCallArguments(forMethod: methodName!, on: obj) else { return }
    b.callComputedMethod(method, on: obj, withArgs: arguments)
}
```
What do you think?

amarekano commented:
Thanks for the feedback, switching from `String` to `Variable` sounds good to me. I've addressed your review comments and updated this PR.

As a side note: this approach has also given me ideas on how to address global var declarations using `StoreToScope`. This is wrt to the [discussion](https://github.com/googleprojectzero/fuzzilli/pull/169#issuecomment-752158233) on the FuzzIL compiler. I'll have a separate PR to run these changes by you.

saelo commented on Sources/Fuzzilli/FuzzIL/Operations.swift:
```diff
@@ -415,6 +415,17 @@ class CallMethod: Operation {
+
+    init(numArguments: Int) {
+        // reference object is the first input and method name is the second input
+        super.init(numInputs: numArguments + 2, numOutputs: 1, attributes: [.isParametric, .isVarargs, .isCall])
```
I think this is now no longer `.isParametric`

saelo replied:
The rough rule of thumbs is that every Operation class that has members other than those already in the `Operation` class is parametric. For example integer values (`LoadInteger`), string values (`LoadProperty` and `CallMethod`), or Arrays (`CallFunctionWithSpread`). Maybe a better definition is that an operation should be parametric if it can/should be mutated by the [OperationMutator](https://github.com/googleprojectzero/fuzzilli/blob/master/Sources/Fuzzilli/Mutators/OperationMutator.swift). The `CallComputedMethod` class now no longer has any members and there is no obvious way in which the OperationMutator could meaningfully mutate it.

amarekano replied:
great, thanks for that explanation that's very helpful. Shall I add this explanation to the comments under Instruction.swift just so that we have this knowledge is captured in the source?

saelo replied:
Yep nice, thanks!

saelo replied:
Now you just need to remove the `.isParametric` here and I'll merge it ;D

amarekano replied:
woops! all done now :)

saelo commented on Sources/Fuzzilli/Lifting/FuzzILLifter.swift:
```diff
@@ -150,6 +150,10 @@ public class FuzzILLifter: Lifter {
 
+        case is CallComputedMethod:
+            let arguments = instr.inputs.dropFirst(2).map({ $0.identifier })
+            w.emit("\(instr.output) <- CallComputedMethod \(input(0)), '\(input(1))', [\(arguments.joined(separator: ", "))]")
```
Nit: please remove the `'` around `input(1)` (similar to `LoadComputedProperty`)

amarekano replied:
done

saelo commented on Sources/Fuzzilli/Minimization/CallArgumentReducer.swift:
```diff
@@ -41,6 +41,15 @@ struct CallArgumentReducer: Reducer {
+                }
+
+                let newOp = CallComputedMethod(numArguments: op.numArguments - 2)
+                let newInstr = Instruction(newOp, output: instr.output, inputs: Array(instr.inputs.dropLast(2)))
```
Why are you dropping two arguments at the same time and not just one like the above cases?

amarekano replied:
That was a mistake on my part. I've corrected it now

saelo commented on Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift:
```diff
@@ -351,6 +351,9 @@ public struct AbstractInterpreter {
             set(instr.output, inferMethodSignature(of: op.methodName, on: instr.input(0)).outputType)
 
+        case is CallComputedMethod:
+            set(instr.output, inferMethodSignature(of: instr.input(1).description, on: instr.input(0)).outputType)
```
This doesn't quite work. Without an additional analyzer that can determine the runtime value of a variable this won't be possible, so I think you can just set the result to `.unknown` directly, similar to how e.g. [`LoadComputedProperty` is handled](https://github.com/googleprojectzero/fuzzilli/blob/40a119bdba15f764b96330c19a3d9de55516459d/Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift#L344). Feel free to make a TODO comment about improving this eventually if you want though

saelo commented on Sources/Fuzzilli/FuzzIL/Instruction.swift:
```diff
@@ -590,6 +592,9 @@ extension Instruction: ProtobufConvertible {
         case .callMethod(let p):
             op = CallMethod(methodName: p.methodName, numArguments: inouts.count - 2)
+        case .callComputedMethod(_):
+            // We subtract 3 from the inouts count since the first two elements are the callee and method and the last element is the instruction index
```
I think this comment isn't correct. The first two are the reference object and the method, and the last element is the output variable (not the instruction index)

saelo commented on Sources/Fuzzilli/Mutators/OperationMutator.swift:
```diff
@@ -79,6 +79,8 @@ public class OperationMutator: BaseInstructionMutator {
             newOp = DeleteElement(index: b.genIndex())
         case let op as CallMethod:
             newOp = CallMethod(methodName: b.genMethodName(), numArguments: op.numArguments)
+        case let op as CallComputedMethod:
```
This no longer mutates anything, so let's remove this (together with the `.isParametric` flag of the operation)

saelo commented on Tests/FuzzilliTests/LifterTest.swift:
```diff
@@ -512,6 +512,29 @@ class LifterTests: XCTestCase {
+
+        let v0 = b.loadString("Hello World")
+        let v2 = b.loadBuiltin("Symbol")
+        let v3 = b.loadProperty("iterator", of: v2)
```
Nit: the variable numbers here don't match the ones in `expected_program` below

saelo commented:
Perfect, thanks!

amarekano commented:
and thank you for your patience with these reviews :D
